File: js/openaiChatCompletion/structuredOutputs.ts
import "dotenv/config";
import { wrapOpenAI } from "langsmith/wrappers";
import { OpenAI } from "openai";
import { zodResponseFormat } from "openai/helpers/zod";
import { z } from "zod";

const client = wrapOpenAI(new OpenAI());

const PersonSchema = z.object({
  name: z.string(),
  age: z.number(),
  occupation: z.string(),
  hobbies: z.array(z.string()),
});

async function main() {
  const response = await client.chat.completions.create({
    model: "gpt-5-2025-08-07",
    messages: [
      {
        role: "user",
        content:
          "Tell me about a fictional software engineer named Alice who is 28 years old.",
      },
    ],
    response_format: zodResponseFormat(PersonSchema, "person"),
  });

  const person = JSON.parse(response.choices[0].message.content!);

  console.log("Structured output response:");
  console.log(person);
}

main();

File: python/wrap_openai_chat/structured_outputs.py
import openai
from langsmith import traceable
from langsmith.wrappers import wrap_openai
from pydantic import BaseModel

client = wrap_openai(openai.Client())


class Person(BaseModel):
    name: str
    age: int
    occupation: str
    hobbies: list[str]


@traceable(name="Chat Completion Structured Output")
def chat_completion_structured_output():
    """Chat completion with structured output."""
    response = client.beta.chat.completions.parse(
        model="gpt-5-2025-08-07",
        messages=[
            {
                "role": "user",
                "content": "Tell me about a fictional software engineer named Alice who is 28 years old.",
            },
        ],
        response_format=Person,
    )

    person = response.choices[0].message.parsed
    print(f"Structured output response: {person}")
    return response


def main():
    print("Running wrap_openai chat completion structured output example...")
    chat_completion_structured_output()
    return {"structured_output": "complete"}
